torch/csrc/jit/passes/break_continue_transform.cpp

```diff
@@ -0,0 +1,326 @@
+#include <torch/csrc/jit/passes/break_continue_transform.h>
+#include <ATen/core/jit_type.h>
+#include <torch/csrc/jit/ir.h>
+#include <torch/csrc/jit/ir_views.h>
+#include <torch/csrc/jit/passes/constant_pooling.h>
+#include <torch/csrc/jit/passes/dead_code_elimination.h>
+#include <torch/csrc/jit/script/error_report.h>
+#include <torch/csrc/jit/script/final_returns.h>
+
+namespace torch {
+namespace jit {
+namespace script {
+
+void moveBlockBeforeNode(Node* before_node, Block* block);
+
+/**
+ * This pass transforms the graph_ so that break & continue statements are
+ * removed. We transform the graph_ so that ops following a break or continue
+ * are not run.
+ */
+
+// Will a block or node continue or break
+enum LoopStatus { WONT, MIGHT, WILL };
+
+// Are we transforming breaks or continues
+enum Transform { BREAKS, CONTINUES };
+
+struct LoopTransformer {
+  LoopTransformer(std::shared_ptr<Graph> graph, Transform transform)
+      : graph_(std::move(graph)) {
+    WithInsertPoint guard(graph_->block()->nodes().front());
+    true_val_ = graph_->insertConstant(true);
+    false_val_ = graph_->insertConstant(false);
+    transform_ = transform;
+  };
+
+  const std::string& getVarname() {
+    static const std::string& break_name = "$did_break";
+    static const std::string& continue_name = "$did_continue";
+    return transform_ == BREAKS ? break_name : continue_name;
+  }
+
+  Symbol transformKind() {
+    return transform_ == BREAKS ? prim::BreakStmt : prim::ContinueStmt;
+  }
+
+  // Recurses on the if node and returns its return status
+  // If status != WONT, sets the block_return_val and sentinel val
+  // of its parent block before exit
+  LoopStatus handleIf(Node* node) {
+    auto true_block = node->blocks().at(0);
+    auto false_block = node->blocks().at(1);
+
+    auto true_status = handleTransforms(true_block);
+    auto false_status = handleTransforms(false_block);
+
+    if (true_status == WONT && false_status == WONT) {
+      return WONT;
+    } else if (true_status == WILL && false_status == WILL) {
+      return WILL;
+    } else {
+      return MIGHT;
+    }
+  }
+
+  LoopStatus guardBlockNodes(
+      Block* block,
+      generic_graph_node_list_iterator<Node>& iter) {
```

Comment:
We have type aliases for this type that is more readable node_list::iterator, maybe? Alternatively just pass the Node*, you can get the iterator from it directly.